llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -3020,91 +3020,111 @@ SelectionDAG::getValidShiftAmountConstant(SDValue V,
     // Shifting more than the bitwidth is not valid.
     const APInt &ShAmt = SA->getAPIntValue();
     if (ShAmt.ult(BitWidth))
-      return &ShAmt;
+      return ShAmt.getZExtValue();
+  } else {
+    KnownBits KnownAmt =
+        computeKnownBits(V.getOperand(1), DemandedElts, Depth + 1);
+    if (KnownAmt.isConstant() && KnownAmt.getConstant().ult(BitWidth))
```

Comment:
IMO if its constant only, Depth should start at MaxDepth - 2, otherwise think there is a lot of wasted compute.

Reply:
The problem we're hitting is that in many cases its not just a bitcast of a constant, its often more complex than that and when trying to use value tracking during legalization we only get one chance, we can't wait for combines to simplify things.

Reply:
Okay, fair enough (maybe mention in a comment? This seems a bit non-intuitive).